### PR TITLE
12085 prevent dupes

### PR DESCRIPTION
**Pull request by jpdimond**

### CODE REVIEW

File: app/models/odk/response_parser.rb
Comment (jpdimond) [reply app/models/odk/response_parser.rb:17]: this is funny because i copied the code that rails is using...

**Review by cooperka — COMMENTED** [review]
thanks for your patience, glad we got to chat about it before I left, finally getting around to full review. looking good, awesome that the checksum generation worked out!

here's where the complexity of implementing this whole feature comes in as opposed to running it ad-hoc on specific missions... I can think of a couple additional cases:
- [x] user submits the same response again after it's been processed earlier (this is handled here nicely)
- [ ] user submits the same response twice at the same time and they both process in different threads so ResponseParser.duplicate? initially returns false for each of them (I don't think this is yet handled, though you do have a spec for threads so maybe you were able to confirm that rails will never put us in this situation? great if so! see related comment in the spec itself)
- [ ] two separate users submit the same answers to a survey but they're really meant to be two independent submissions (this is a new concern I came up with, see below)

to think about mitigation for the 3rd case, I ran `ActiveStorage::Attachment.where(record_type: "Response").last.download` on the staging server and got an answer that looks like this:

```xml
<?xml version='1.0' ?>
<data id="3106cd6e-b043-4723-8e33-97474555f630" version="2021032200" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <qingd154d4e8-6566-42bd-98f5-6ffcabd26c8c>1643311886219.jpg</qingd154d4e8-6566-42bd-98f5-6ffcabd26c8c>
</data>
```

it's devoid of context other than the form version and list of answers, so in order to deduplicate properly I think we'll also need to take more info into account -- likely either the submitter user ID, device ID, and/or IP address. my preference is to use the submitter user ID because the others seem less useful, let's chat soon and confirm with dottie.

File: app/models/odk/response_parser.rb
Comment (jpdimond) [reply app/models/odk/response_parser.rb:16]: cool, will do

File: spec/requests/odk/submission_spec.rb
Comment (jpdimond) [reply spec/requests/odk/submission_spec.rb:190]: ok, sooooo i was thinking in real life conditions that the server would not receive the request as fast as the test framework could make the threads. but maybe i'm wrong. this does fail if the sleep is not in there because we have a race condition where the response has not been saved yet, so we can't look at the checksum yet. i leave it up to you as to whether we should continue to problem solve around it. we could consider locking around user id or something?

Comment (jpdimond) [reply spec/requests/odk/submission_spec.rb:190]: ok now it is not failing....

File: spec/models/odk/response_parser_spec.rb
Comment (jpdimond) [reply spec/models/odk/response_parser_spec.rb:551]: done

File: spec/requests/odk/submission_spec.rb
Comment (jpdimond) [reply spec/requests/odk/submission_spec.rb:175]: ah yes, thx!

File: spec/requests/odk/submission_spec.rb
Comment (jpdimond) [reply spec/requests/odk/submission_spec.rb:190]: i thought there was a race condition but there doesn't seem to be now! which makes me a lil hmm. i was also thinking we could have a unique constraint on the checksum on the database level so even with race conditions it wouldn't happen. but then we would need to address the case that another user has an exact same submission as you mentioned.

Comment (cooperka) [reply spec/requests/odk/submission_spec.rb]: hmm yes unless we can explicitly confirm that rails never parallelizes these requests (I think it does parallelize them), I think we need to address checkbox #2 above to handle this.

imagined scenario:
* user submits a big response that takes 120 seconds to process
* 60 seconds later, user gets bored, clicks back in the app, and tries submitting again
* rails accepts the second request, starts a parallel thread, verifies that no checksum matches yet, and works on saving it
* we end up with a duplicate

imagined mitigation strategy:
* at the very end when saving the response, double check that it's STILL not a duplicate; if it is determined to NOW be a duplicate, we need to destroy it (this logic will be safe and won't run in parallel, Rails guarantees that)
* this spec SHOULD fail until this new check exists

this new check could indeed come in the form of a uniqueness constraint such as unique [checksum, submitter ID] -- but that would require cleaning up legacy data which we aren't doing (many things would be easier if we cleaned up legacy data but I understand why it's important not to for research/consistency reasons).

so instead I think this just needs to be regular code to basically run your `duplicate?` check again at the end and if detected, destroy the duplicate instead of saving it.

File: spec/requests/odk/submission_spec.rb
Comment (jpdimond) [reply spec/requests/odk/submission_spec.rb:178]: done

File: spec/models/odk/response_saver_spec.rb
Comment (jpdimond) [inline spec/models/odk/response_saver_spec.rb:2]: This test simulates this:
image

File: spec/models/odk/response_saver_spec.rb
Comment (jpdimond) [inline spec/models/odk/response_saver_spec.rb:66]: There isn't a way of simulating a new rails workers, but you can simulate a new database thread which accomplishes the same thing.

Comment (jpdimond) [reply spec/models/odk/response_saver_spec.rb:66]: This method also just inserts the blob, and doesn't insert the response because we are testing duplicates on the checksum which is on the blob.

File: app/models/odk/response_saver.rb
Comment (jpdimond) [reply app/models/odk/response_saver.rb:8]: blerf

**Review by cooperka — COMMENTED** [review]
phew this is one of the trickiest things I've dealt with in a long time, I'm excited that you got all this working with relatively concise code overall -- seems like this will be no problem to maintain now that we have these specs working.

tested and verified with odk collect, using a debugger breakpoint within `ODK::ResponseParser#populate_response` to submit twice and ignore the duplicate. YES!!

oh, we should log to Sentry too, to see if one or the other of these checks gets triggered more often. I've added `Sentry.capture_message` ([docs](https://docs.sentry.io/platforms/ruby/usage/)) in a couple places ✔️ 

just a couple of small comments left, excited to get this into the world next week 🎉

File: app/models/odk/response_saver.rb
Comment (jpdimond) [reply app/models/odk/response_saver.rb:6]: When reading up on postgres and isolation levels, it was recommended to add retries. I saw 10 in a couple places e.g. https://franckpachot.medium.com/pgbench-retry-for-repeatable-read-transactions-first-re-tries-5c3795bc5d68
so happy to change it to that, but I was concerned about tryinig too much and bogging down the server, so 5 seemed safer?

File: app/models/odk/response_saver.rb
Comment (jpdimond) [reply app/models/odk/response_saver.rb:8]: haha ty!!!

File: app/models/odk/response_parser.rb
Comment (jpdimond) [reply app/models/odk/response_parser.rb:13]: good catch, done!

File: app/models/odk/response_saver.rb
Comment (jpdimond) [reply app/models/odk/response_saver.rb]: actually i changed it to 10, i don't wanna lose data!

File: app/models/odk/response_parser.rb
Comment (cooperka) [inline app/models/odk/response_parser.rb:18]: Ah this was just backwards fyi, should be `&.` instead of `.&`

If helpful for memory jogging, here's the rationale behind this horrible syntax because it doesn't look like most other languages: https://stackoverflow.com/a/36813420/763231 ... you can think of it similar to `response && .user && .id` where it doesn't continue unless the preceding thing exists.

Fine to leave as-is!

File: app/models/odk/response_saver.rb
Comment (cooperka) [reply app/models/odk/response_saver.rb]: gotcha, either seems okay 👍  it seems like we wouldn't lose data in any case, just return an error status to Collect and require re-submission. thanks!

I'm gonna add similar Sentry logging here so we detect whenever this condition is being triggered. it's possible this retry will end up being a bottleneck, but I doubt it -- Scout can tell us if there ends up being an issue

**Review by cooperka — APPROVED** [review]
🎉🎉🎉

**Comment by cooperka** [comment]
[rebased on main to clean up 7 duplicate commits that somehow got inserted out of order, verified the code is identical, letting specs run one last time before merging; screenshots below for posterity]

12085 before
12085 after

**Pull request merged**
